// @zombienet/orchestrator@0.0.68/dist/providers/podman/resources/bootnodeResource.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BootNodeResource = void 0;
const sharedTypes_1 = require("../../../sharedTypes");
const nodeResource_1 = require("./nodeResource");
class BootNodeResource extends nodeResource_1.NodeResource {
    constructor(client, namespace, nodeSetupConfig) {
        super(client, namespace, nodeSetupConfig);
    }
    generatePodSpec(containers, volumes) {
        return {
            apiVersion: "v1",
            kind: "Pod",
            metadata: {
                name: "bootnode",
                namespace: this.namespace,
                labels: {
                    "zombie-role": sharedTypes_1.ZombieRole.BootNode,
                    app: "zombienet",
                    "zombie-ns": this.namespace,
                },
            },
            spec: {
                hostname: "bootnode",
                initContainers: [],
                restartPolicy: "OnFailure",
                volumes,
                containers,
            },
        };
    }
}
exports.BootNodeResource = BootNodeResource;
