// @zombienet/orchestrator@0.0.68/dist/providers/k8s/resources/bootNodeResource.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BootNodeResource = void 0;
const sharedTypes_1 = require("../../../sharedTypes");
const nodeResource_1 = require("./nodeResource");
class BootNodeResource extends nodeResource_1.NodeResource {
    constructor(namespace, nodeSetupConfig) {
        super(namespace, nodeSetupConfig);
    }
    generatePodSpec(initContainers, containers, volumes) {
        return {
            apiVersion: "v1",
            kind: "Pod",
            metadata: {
                name: "bootnode",
                labels: {
                    "app.kubernetes.io/name": this.namespace,
                    "app.kubernetes.io/instance": "bootnode",
                    "zombie-role": sharedTypes_1.ZombieRole.BootNode,
                    app: "zombienet",
                    "x-infra-instance": process.env.X_INFRA_INSTANCE || "ondemand",
                },
            },
            spec: {
                hostname: "bootnode",
                containers,
                initContainers,
                restartPolicy: "Always",
                volumes,
                securityContext: {
                    fsGroup: 1000,
                    runAsUser: 1000,
                    runAsGroup: 1000,
                },
            },
        };
    }
}
exports.BootNodeResource = BootNodeResource;
