// @moonwall/util@4.4.4/dist/chunk-MPBUERDP.js
// src/constants/chain.ts
import "@moonbeam-network/api-augment";
var SPECS_PATH = `./moonbeam-test-specs`;
var DEBUG_MODE = process.env.DEBUG_MODE || false;
var DISPLAY_LOG = process.env.MOONBEAM_LOG || false;
var MOONBEAM_LOG = process.env.MOONBEAM_LOG || "info";
var BASE_PATH = process.env.BASE_PATH;
var CUSTOM_SPEC_PATH = process.env.CUSTOM_SPEC_PATH;
var BINARY_PATH = process.env.BINARY_PATH || `../target/release/moonbeam`;
var RELAY_BINARY_PATH = process.env.RELAY_BINARY_PATH || `../target/release/polkadot`;
var RELAY_LOG = process.env.RELAY_LOG;
var OVERRIDE_RUNTIME_PATH = process.env["OVERRIDE_RUNTIME_PATH"] || void 0;
var SPAWNING_TIME = 2e4;
var ETHAPI_CMD = process.env.ETHAPI_CMD || "";
var WASM_RUNTIME_OVERRIDES = process.env.WASM_RUNTIME_OVERRIDES || "";
var RELAY_CHAIN_NODE_NAMES = ["Alice", "Bob", "Charlie", "Dave", "Eve", "Ferdie", "One"];
var ZERO_ADDRESS = "0x0000000000000000000000000000000000000000";
var TREASURY_ACCOUNT = "0x6d6f646c70632f74727372790000000000000000";
var GLMR = 1000000000000000000n;
var MILLIGLMR = 1000000000000000n;
var MICROGLMR = 1000000000000n;
var DEFAULT_GENESIS_BALANCE = 2n ** 80n;
var DEFAULT_GENESIS_STAKING = 1000n * GLMR;
var DEFAULT_GENESIS_MAPPING = 100n * GLMR;
var PROPOSAL_AMOUNT = 1000n * GLMR;
var VOTE_AMOUNT = 10n * GLMR;
var MIN_GLMR_STAKING = 1000n * GLMR;
var MIN_GLMR_DELEGATOR = 1n * GLMR;
var WEIGHT_PER_SECOND = 1000000000000n;
var GAS_PER_SECOND = 40000000n;
var GAS_PER_WEIGHT = WEIGHT_PER_SECOND / GAS_PER_SECOND;
var GAS_PER_POV_BYTES = 4n;
var BLOCK_WEIGHT_LIMIT = WEIGHT_PER_SECOND / 2n;
var BLOCK_GAS_LIMIT = BLOCK_WEIGHT_LIMIT / GAS_PER_WEIGHT;
var EXTRINSIC_GAS_LIMIT = BLOCK_GAS_LIMIT * 3n / 4n - BLOCK_GAS_LIMIT / 10n;
var MAX_BLOCK_DEV_POV = 4 * 1024 * 1024 + 512;
var MAX_ETH_POV_PER_TX = EXTRINSIC_GAS_LIMIT / GAS_PER_POV_BYTES;
var EXTRINSIC_BASE_WEIGHT = 25e7;
var WEIGHT_PER_GAS = 1000000000000n / 40000000n;
var MIN_GAS_PRICE = 10000000000n;
var GAS_LIMIT_POV_RATIO = 4;
var PRECOMPILE_PARACHAIN_STAKING_ADDRESS = "0x0000000000000000000000000000000000000800";
var PRECOMPILE_CROWDLOAN_REWARDS_ADDRESS = "0x0000000000000000000000000000000000000801";
var PRECOMPILE_NATIVE_ERC20_ADDRESS = "0x0000000000000000000000000000000000000802";
var PRECOMPILE_DEMOCRACY_ADDRESS = "0x0000000000000000000000000000000000000803";
var PRECOMPILE_XTOKENS_ADDRESS = "0x0000000000000000000000000000000000000804";
var PRECOMPILE_RELAY_ENCODER_ADDRESS = "0x0000000000000000000000000000000000000805";
var PRECOMPILE_XCM_TRANSACTOR_ADDRESS_V1 = "0x0000000000000000000000000000000000000806";
var PRECOMPILE_AUTHOR_MAPPING_ADDRESS = "0x0000000000000000000000000000000000000807";
var PRECOMPILE_BATCH_ADDRESS = "0x0000000000000000000000000000000000000808";
var PRECOMPILE_RANDOMNESS_ADDRESS = "0x0000000000000000000000000000000000000809";
var PRECOMPILE_CALL_PERMIT_ADDRESS = "0x000000000000000000000000000000000000080a";
var PRECOMPILE_PROXY_ADDRESS = "0x000000000000000000000000000000000000080b";
var PRECOMPILE_XCM_UTILS_ADDRESS = "0x000000000000000000000000000000000000080c";
var PRECOMPILE_XCM_TRANSACTOR_ADDRESS_V2 = "0x000000000000000000000000000000000000080d";
var PRECOMPILE_COUNCIL_ADDRESS = "0x000000000000000000000000000000000000080e";
var PRECOMPILE_TECH_COMMITTEE_ADDRESS = "0x000000000000000000000000000000000000080f";
var PRECOMPILE_TREASURY_COUNCIL_ADDRESS = "0x0000000000000000000000000000000000000810";
var PRECOMPILE_DISPATCH_ADDRESS = "0x0000000000000000000000000000000000000401";
var PRECOMPILE_REFERENDA_ADDRESS = "0x0000000000000000000000000000000000000811";
var PRECOMPILE_CONVICTION_VOTING_ADDRESS = "0x0000000000000000000000000000000000000812";
var PRECOMPILE_PREIMAGE_ADDRESS = "0x0000000000000000000000000000000000000813";
var PRECOMPILE_OPEN_TECH_COMMITTEE_ADDRESS = "0x0000000000000000000000000000000000000814";
var PRECOMPILE_GMP_ADDRESS = "0x0000000000000000000000000000000000000816";
var PRECOMPILES = {
  ParachainStaking: "0x0000000000000000000000000000000000000800",
  CrowdloanRewards: "0x0000000000000000000000000000000000000801",
  NativeErc20: ["0x0000000000000000000000000000000000000802", "IERC20"],
  Democracy: "0x0000000000000000000000000000000000000803",
  Xtokens: "0x0000000000000000000000000000000000000804",
  RelayEncoder: "0x0000000000000000000000000000000000000805",
  XcmTransactorV1: "0x0000000000000000000000000000000000000806",
  AuthorMapping: "0x0000000000000000000000000000000000000807",
  Batch: "0x0000000000000000000000000000000000000808",
  Randomness: "0x0000000000000000000000000000000000000809",
  CallPermit: "0x000000000000000000000000000000000000080a",
  Proxy: "0x000000000000000000000000000000000000080b",
  XcmUtils: "0x000000000000000000000000000000000000080c",
  XcmTransactorV2: "0x000000000000000000000000000000000000080d",
  Council: ["0x000000000000000000000000000000000000080e", "Collective"],
  TechCommittee: ["0x000000000000000000000000000000000000080f", "Collective"],
  TreasuryCouncil: ["0x0000000000000000000000000000000000000810", "Collective"],
  // DISPATCH: "0x0000000000000000000000000000000000000401",
  Referenda: "0x0000000000000000000000000000000000000811",
  ConvictionVoting: "0x0000000000000000000000000000000000000812",
  Preimage: "0x0000000000000000000000000000000000000813",
  // OpenTechCommittee: "0x0000000000000000000000000000000000000814",
  Gmp: "0x0000000000000000000000000000000000000816"
};
var MINUTES = 60 / 12;
var HOURS = MINUTES * 60;
var DAYS = HOURS * 24;
var TWO_MINS = 2 * 60 * 1e3;
var FIVE_MINS = 5 * 60 * 1e3;
var TEN_MINS = 10 * 60 * 1e3;
var THIRTY_MINS = 30 * 60 * 1e3;
var ONE_HOURS = 60 * 60 * 1e3;
var TWO_HOURS = 2 * 60 * 60 * 1e3;
var THREE_HOURS = 3 * 60 * 60 * 1e3;
var FOUR_HOURS = 4 * 60 * 60 * 1e3;
var CONTRACT_RANDOMNESS_STATUS_DOES_NOT_EXISTS = 0;
var CONTRACT_RANDOMNESS_STATUS_PENDING = 1;
var CONTRACT_RANDOMNESS_STATUS_READY = 2;
var CONTRACT_RANDOMNESS_STATUS_EXPIRED = 3;
var CONTRACT_PROXY_TYPE_ANY = 0;
var CONTRACT_PROXY_TYPE_NON_TRANSFER = 1;
var CONTRACT_PROXY_TYPE_GOVERNANCE = 2;
var CONTRACT_PROXY_TYPE_STAKING = 3;
var CONTRACT_PROXY_TYPE_CANCEL_PROXY = 4;
var CONTRACT_PROXY_TYPE_BALANCES = 5;
var CONTRACT_PROXY_TYPE_AUTHOR_MAPPING = 6;
var CONTRACT_PROXY_TYPE_IDENTITY_JUDGEMENT = 7;
var MAX_BLOCK_WEIGHT = 5e11;
var TARGET_FILL_AMOUNT = (MAX_BLOCK_WEIGHT * 0.75 * 0.25 - EXTRINSIC_BASE_WEIGHT) / MAX_BLOCK_WEIGHT * 1e9;
var TARGET_FILL_PERMILL = 0.25 * 1e6;
var WEIGHT_FEE = 50n * 1000n;
var RUNTIME_CONSTANTS = {
  MOONBASE: {
    MIN_FEE_MULTIPLIER: "100000000000000000",
    MAX_FEE_MULTIPLIER: "100000000000000000000000",
    MIN_BASE_FEE_IN_WEI: "125000000",
    MAX_BASE_FEE_IN_WEI: "125000000000000"
  },
  MOONRIVER: {
    MIN_FEE_MULTIPLIER: "1000000000000000000",
    MAX_FEE_MULTIPLIER: "100000000000000000000000",
    MIN_BASE_FEE_IN_WEI: "1250000000",
    MAX_BASE_FEE_IN_WEI: "125000000000000"
  },
  MOONBEAM: {
    MIN_FEE_MULTIPLIER: "1000000000000000000",
    MAX_FEE_MULTIPLIER: "100000000000000000000000",
    MIN_BASE_FEE_IN_WEI: "125000000000",
    MAX_BASE_FEE_IN_WEI: "12500000000000000"
  }
};
var DUMMY_REVERT_BYTECODE = "0x60006000fd";

export {
  SPECS_PATH,
  DEBUG_MODE,
  DISPLAY_LOG,
  MOONBEAM_LOG,
  BASE_PATH,
  CUSTOM_SPEC_PATH,
  BINARY_PATH,
  RELAY_BINARY_PATH,
  RELAY_LOG,
  OVERRIDE_RUNTIME_PATH,
  SPAWNING_TIME,
  ETHAPI_CMD,
  WASM_RUNTIME_OVERRIDES,
  RELAY_CHAIN_NODE_NAMES,
  ZERO_ADDRESS,
  TREASURY_ACCOUNT,
  GLMR,
  MILLIGLMR,
  MICROGLMR,
  DEFAULT_GENESIS_BALANCE,
  DEFAULT_GENESIS_STAKING,
  DEFAULT_GENESIS_MAPPING,
  PROPOSAL_AMOUNT,
  VOTE_AMOUNT,
  MIN_GLMR_STAKING,
  MIN_GLMR_DELEGATOR,
  WEIGHT_PER_SECOND,
  GAS_PER_SECOND,
  GAS_PER_WEIGHT,
  GAS_PER_POV_BYTES,
  BLOCK_WEIGHT_LIMIT,
  BLOCK_GAS_LIMIT,
  EXTRINSIC_GAS_LIMIT,
  MAX_BLOCK_DEV_POV,
  MAX_ETH_POV_PER_TX,
  EXTRINSIC_BASE_WEIGHT,
  WEIGHT_PER_GAS,
  MIN_GAS_PRICE,
  GAS_LIMIT_POV_RATIO,
  PRECOMPILE_PARACHAIN_STAKING_ADDRESS,
  PRECOMPILE_CROWDLOAN_REWARDS_ADDRESS,
  PRECOMPILE_NATIVE_ERC20_ADDRESS,
  PRECOMPILE_DEMOCRACY_ADDRESS,
  PRECOMPILE_XTOKENS_ADDRESS,
  PRECOMPILE_RELAY_ENCODER_ADDRESS,
  PRECOMPILE_XCM_TRANSACTOR_ADDRESS_V1,
  PRECOMPILE_AUTHOR_MAPPING_ADDRESS,
  PRECOMPILE_BATCH_ADDRESS,
  PRECOMPILE_RANDOMNESS_ADDRESS,
  PRECOMPILE_CALL_PERMIT_ADDRESS,
  PRECOMPILE_PROXY_ADDRESS,
  PRECOMPILE_XCM_UTILS_ADDRESS,
  PRECOMPILE_XCM_TRANSACTOR_ADDRESS_V2,
  PRECOMPILE_COUNCIL_ADDRESS,
  PRECOMPILE_TECH_COMMITTEE_ADDRESS,
  PRECOMPILE_TREASURY_COUNCIL_ADDRESS,
  PRECOMPILE_DISPATCH_ADDRESS,
  PRECOMPILE_REFERENDA_ADDRESS,
  PRECOMPILE_CONVICTION_VOTING_ADDRESS,
  PRECOMPILE_PREIMAGE_ADDRESS,
  PRECOMPILE_OPEN_TECH_COMMITTEE_ADDRESS,
  PRECOMPILE_GMP_ADDRESS,
  PRECOMPILES,
  MINUTES,
  HOURS,
  DAYS,
  TWO_MINS,
  FIVE_MINS,
  TEN_MINS,
  THIRTY_MINS,
  ONE_HOURS,
  TWO_HOURS,
  THREE_HOURS,
  FOUR_HOURS,
  CONTRACT_RANDOMNESS_STATUS_DOES_NOT_EXISTS,
  CONTRACT_RANDOMNESS_STATUS_PENDING,
  CONTRACT_RANDOMNESS_STATUS_READY,
  CONTRACT_RANDOMNESS_STATUS_EXPIRED,
  CONTRACT_PROXY_TYPE_ANY,
  CONTRACT_PROXY_TYPE_NON_TRANSFER,
  CONTRACT_PROXY_TYPE_GOVERNANCE,
  CONTRACT_PROXY_TYPE_STAKING,
  CONTRACT_PROXY_TYPE_CANCEL_PROXY,
  CONTRACT_PROXY_TYPE_BALANCES,
  CONTRACT_PROXY_TYPE_AUTHOR_MAPPING,
  CONTRACT_PROXY_TYPE_IDENTITY_JUDGEMENT,
  MAX_BLOCK_WEIGHT,
  TARGET_FILL_AMOUNT,
  TARGET_FILL_PERMILL,
  WEIGHT_FEE,
  RUNTIME_CONSTANTS,
  DUMMY_REVERT_BYTECODE
};
