// @acala-network/chopsticks@0.9.3/dist/cjs/utils/decoder.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "decodeStorageDiff", {
    enumerable: true,
    get: function() {
        return decodeStorageDiff;
    }
});
const _chopstickscore = require("@acala-network/chopsticks-core");
const _jsondiffpatch = require("jsondiffpatch");
const _lodash = /*#__PURE__*/ _interop_require_default(require("lodash"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const diffPatcher = (0, _jsondiffpatch.create)({
    array: {
        detectMove: false
    },
    textDiff: {
        minLength: Number.MAX_VALUE
    }
});
const decodeStorageDiff = async (block, diff)=>{
    const [oldState, newState] = await (0, _chopstickscore.decodeBlockStorageDiff)(block, diff);
    const oldStateWithoutEvents = _lodash.default.cloneDeep(oldState);
    if (oldStateWithoutEvents['system']?.['events']) {
        oldStateWithoutEvents['system']['events'] = [];
    }
    return {
        oldState,
        newState,
        delta: diffPatcher.diff(oldStateWithoutEvents, newState)
    };
};
