// @acala-network/chopsticks-core@0.9.3/dist/cjs/rpc/substrate/system.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    system_accountNextIndex: function() {
        return system_accountNextIndex;
    },
    system_chain: function() {
        return system_chain;
    },
    system_chainType: function() {
        return system_chainType;
    },
    system_dryRun: function() {
        return system_dryRun;
    },
    system_health: function() {
        return system_health;
    },
    system_localListenAddresses: function() {
        return system_localListenAddresses;
    },
    system_localPeerId: function() {
        return system_localPeerId;
    },
    system_name: function() {
        return system_name;
    },
    system_nodeRoles: function() {
        return system_nodeRoles;
    },
    system_properties: function() {
        return system_properties;
    },
    system_version: function() {
        return system_version;
    }
});
const _util = require("@polkadot/util");
const system_localPeerId = async ()=>'5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY';
const system_nodeRoles = async ()=>[
        'Full'
    ];
const system_localListenAddresses = async ()=>[];
const system_chain = async (context)=>{
    return context.chain.api.getSystemChain();
};
const system_properties = async (context)=>{
    return context.chain.api.getSystemProperties();
};
const system_name = async (context)=>{
    return context.chain.api.getSystemName();
};
const system_version = async (_context)=>{
    return 'chopsticks-v1';
};
const system_chainType = async (_context)=>{
    return 'Development';
};
const system_health = async ()=>{
    return {
        peers: 0,
        isSyncing: false,
        shouldHavePeers: false
    };
};
const system_dryRun = async (context, [extrinsic, at])=>{
    const { outcome } = await context.chain.dryRunExtrinsic(extrinsic, at);
    return outcome.toHex();
};
const system_accountNextIndex = async (context, [address])=>{
    const head = context.chain.head;
    const registry = await head.registry;
    const account = registry.createType('AccountId', address);
    const result = await head.call('AccountNonceApi_account_nonce', [
        account.toHex()
    ]);
    const nonce = registry.createType('Index', (0, _util.hexToU8a)(result.result)).toNumber();
    return nonce + context.chain.txPool.pendingExtrinsicsBy(address).length;
};
