// @moonwall/util@4.4.4/dist/constants/chain.js
import {
  BASE_PATH,
  BINARY_PATH,
  BLOCK_GAS_LIMIT,
  BLOCK_WEIGHT_LIMIT,
  CONTRACT_PROXY_TYPE_ANY,
  CONTRACT_PROXY_TYPE_AUTHOR_MAPPING,
  CONTRACT_PROXY_TYPE_BALANCES,
  CONTRACT_PROXY_TYPE_CANCEL_PROXY,
  CONTRACT_PROXY_TYPE_GOVERNANCE,
  CONTRACT_PROXY_TYPE_IDENTITY_JUDGEMENT,
  CONTRACT_PROXY_TYPE_NON_TRANSFER,
  CONTRACT_PROXY_TYPE_STAKING,
  CONTRACT_RANDOMNESS_STATUS_DOES_NOT_EXISTS,
  CONTRACT_RANDOMNESS_STATUS_EXPIRED,
  CONTRACT_RANDOMNESS_STATUS_PENDING,
  CONTRACT_RANDOMNESS_STATUS_READY,
  CUSTOM_SPEC_PATH,
  DAYS,
  DEBUG_MODE,
  DEFAULT_GENESIS_BALANCE,
  DEFAULT_GENESIS_MAPPING,
  DEFAULT_GENESIS_STAKING,
  DISPLAY_LOG,
  DUMMY_REVERT_BYTECODE,
  ETHAPI_CMD,
  EXTRINSIC_BASE_WEIGHT,
  EXTRINSIC_GAS_LIMIT,
  FIVE_MINS,
  FOUR_HOURS,
  GAS_LIMIT_POV_RATIO,
  GAS_PER_POV_BYTES,
  GAS_PER_SECOND,
  GAS_PER_WEIGHT,
  GLMR,
  HOURS,
  MAX_BLOCK_DEV_POV,
  MAX_BLOCK_WEIGHT,
  MAX_ETH_POV_PER_TX,
  MICROGLMR,
  MILLIGLMR,
  MINUTES,
  MIN_GAS_PRICE,
  MIN_GLMR_DELEGATOR,
  MIN_GLMR_STAKING,
  MOONBEAM_LOG,
  ONE_HOURS,
  OVERRIDE_RUNTIME_PATH,
  PRECOMPILES,
  PRECOMPILE_AUTHOR_MAPPING_ADDRESS,
  PRECOMPILE_BATCH_ADDRESS,
  PRECOMPILE_CALL_PERMIT_ADDRESS,
  PRECOMPILE_CONVICTION_VOTING_ADDRESS,
  PRECOMPILE_COUNCIL_ADDRESS,
  PRECOMPILE_CROWDLOAN_REWARDS_ADDRESS,
  PRECOMPILE_DEMOCRACY_ADDRESS,
  PRECOMPILE_DISPATCH_ADDRESS,
  PRECOMPILE_GMP_ADDRESS,
  PRECOMPILE_NATIVE_ERC20_ADDRESS,
  PRECOMPILE_OPEN_TECH_COMMITTEE_ADDRESS,
  PRECOMPILE_PARACHAIN_STAKING_ADDRESS,
  PRECOMPILE_PREIMAGE_ADDRESS,
  PRECOMPILE_PROXY_ADDRESS,
  PRECOMPILE_RANDOMNESS_ADDRESS,
  PRECOMPILE_REFERENDA_ADDRESS,
  PRECOMPILE_RELAY_ENCODER_ADDRESS,
  PRECOMPILE_TECH_COMMITTEE_ADDRESS,
  PRECOMPILE_TREASURY_COUNCIL_ADDRESS,
  PRECOMPILE_XCM_TRANSACTOR_ADDRESS_V1,
  PRECOMPILE_XCM_TRANSACTOR_ADDRESS_V2,
  PRECOMPILE_XCM_UTILS_ADDRESS,
  PRECOMPILE_XTOKENS_ADDRESS,
  PROPOSAL_AMOUNT,
  RELAY_BINARY_PATH,
  RELAY_CHAIN_NODE_NAMES,
  RELAY_LOG,
  RUNTIME_CONSTANTS,
  SPAWNING_TIME,
  SPECS_PATH,
  TARGET_FILL_AMOUNT,
  TARGET_FILL_PERMILL,
  TEN_MINS,
  THIRTY_MINS,
  THREE_HOURS,
  TREASURY_ACCOUNT,
  TWO_HOURS,
  TWO_MINS,
  VOTE_AMOUNT,
  WASM_RUNTIME_OVERRIDES,
  WEIGHT_FEE,
  WEIGHT_PER_GAS,
  WEIGHT_PER_SECOND,
  ZERO_ADDRESS
} from "../chunk-MPBUERDP.js";
export {
  BASE_PATH,
  BINARY_PATH,
  BLOCK_GAS_LIMIT,
  BLOCK_WEIGHT_LIMIT,
  CONTRACT_PROXY_TYPE_ANY,
  CONTRACT_PROXY_TYPE_AUTHOR_MAPPING,
  CONTRACT_PROXY_TYPE_BALANCES,
  CONTRACT_PROXY_TYPE_CANCEL_PROXY,
  CONTRACT_PROXY_TYPE_GOVERNANCE,
  CONTRACT_PROXY_TYPE_IDENTITY_JUDGEMENT,
  CONTRACT_PROXY_TYPE_NON_TRANSFER,
  CONTRACT_PROXY_TYPE_STAKING,
  CONTRACT_RANDOMNESS_STATUS_DOES_NOT_EXISTS,
  CONTRACT_RANDOMNESS_STATUS_EXPIRED,
  CONTRACT_RANDOMNESS_STATUS_PENDING,
  CONTRACT_RANDOMNESS_STATUS_READY,
  CUSTOM_SPEC_PATH,
  DAYS,
  DEBUG_MODE,
  DEFAULT_GENESIS_BALANCE,
  DEFAULT_GENESIS_MAPPING,
  DEFAULT_GENESIS_STAKING,
  DISPLAY_LOG,
  DUMMY_REVERT_BYTECODE,
  ETHAPI_CMD,
  EXTRINSIC_BASE_WEIGHT,
  EXTRINSIC_GAS_LIMIT,
  FIVE_MINS,
  FOUR_HOURS,
  GAS_LIMIT_POV_RATIO,
  GAS_PER_POV_BYTES,
  GAS_PER_SECOND,
  GAS_PER_WEIGHT,
  GLMR,
  HOURS,
  MAX_BLOCK_DEV_POV,
  MAX_BLOCK_WEIGHT,
  MAX_ETH_POV_PER_TX,
  MICROGLMR,
  MILLIGLMR,
  MINUTES,
  MIN_GAS_PRICE,
  MIN_GLMR_DELEGATOR,
  MIN_GLMR_STAKING,
  MOONBEAM_LOG,
  ONE_HOURS,
  OVERRIDE_RUNTIME_PATH,
  PRECOMPILES,
  PRECOMPILE_AUTHOR_MAPPING_ADDRESS,
  PRECOMPILE_BATCH_ADDRESS,
  PRECOMPILE_CALL_PERMIT_ADDRESS,
  PRECOMPILE_CONVICTION_VOTING_ADDRESS,
  PRECOMPILE_COUNCIL_ADDRESS,
  PRECOMPILE_CROWDLOAN_REWARDS_ADDRESS,
  PRECOMPILE_DEMOCRACY_ADDRESS,
  PRECOMPILE_DISPATCH_ADDRESS,
  PRECOMPILE_GMP_ADDRESS,
  PRECOMPILE_NATIVE_ERC20_ADDRESS,
  PRECOMPILE_OPEN_TECH_COMMITTEE_ADDRESS,
  PRECOMPILE_PARACHAIN_STAKING_ADDRESS,
  PRECOMPILE_PREIMAGE_ADDRESS,
  PRECOMPILE_PROXY_ADDRESS,
  PRECOMPILE_RANDOMNESS_ADDRESS,
  PRECOMPILE_REFERENDA_ADDRESS,
  PRECOMPILE_RELAY_ENCODER_ADDRESS,
  PRECOMPILE_TECH_COMMITTEE_ADDRESS,
  PRECOMPILE_TREASURY_COUNCIL_ADDRESS,
  PRECOMPILE_XCM_TRANSACTOR_ADDRESS_V1,
  PRECOMPILE_XCM_TRANSACTOR_ADDRESS_V2,
  PRECOMPILE_XCM_UTILS_ADDRESS,
  PRECOMPILE_XTOKENS_ADDRESS,
  PROPOSAL_AMOUNT,
  RELAY_BINARY_PATH,
  RELAY_CHAIN_NODE_NAMES,
  RELAY_LOG,
  RUNTIME_CONSTANTS,
  SPAWNING_TIME,
  SPECS_PATH,
  TARGET_FILL_AMOUNT,
  TARGET_FILL_PERMILL,
  TEN_MINS,
  THIRTY_MINS,
  THREE_HOURS,
  TREASURY_ACCOUNT,
  TWO_HOURS,
  TWO_MINS,
  VOTE_AMOUNT,
  WASM_RUNTIME_OVERRIDES,
  WEIGHT_FEE,
  WEIGHT_PER_GAS,
  WEIGHT_PER_SECOND,
  ZERO_ADDRESS
};
