// @acala-network/chopsticks-db@0.9.3/dist/cjs/db/entities.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    BlockEntity: function() {
        return BlockEntity;
    },
    KeyValuePair: function() {
        return KeyValuePair;
    }
});
const _typeorm = require("typeorm");
const KeyValuePair = new _typeorm.EntitySchema({
    name: 'KeyValuePair',
    columns: {
        blockHash: {
            primary: true,
            type: 'varchar',
            nullable: false
        },
        key: {
            primary: true,
            type: 'varchar',
            nullable: false
        },
        value: {
            type: 'text',
            nullable: true
        }
    }
});
const BlockEntity = new _typeorm.EntitySchema({
    name: 'Block',
    columns: {
        hash: {
            primary: true,
            type: 'varchar',
            nullable: false
        },
        number: {
            type: 'int',
            nullable: false
        },
        header: {
            type: 'text',
            nullable: false
        },
        parentHash: {
            type: 'varchar',
            nullable: true
        },
        extrinsics: {
            type: 'simple-array',
            nullable: false
        },
        storageDiff: {
            type: 'simple-json',
            nullable: true
        }
    }
});
