// @acala-network/chopsticks@0.9.3/dist/cjs/plugins/dry-run/cli.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "cli", {
    enumerable: true,
    get: function() {
        return cli;
    }
});
const _clioptions = require("../../cli-options.js");
const _dryrunextrinsic = require("./dry-run-extrinsic.js");
const _dryrunpreimage = require("./dry-run-preimage.js");
const cli = (y)=>{
    y.command('dry-run', 'Dry run an extrinsic', (yargs)=>yargs.options({
            ..._clioptions.defaultOptions,
            ..._clioptions.mockOptions,
            extrinsic: {
                desc: 'Extrinsic or call to dry run. If you pass call here then address is required to fake signature',
                string: true
            },
            address: {
                desc: 'Address to fake sign extrinsic',
                string: true
            },
            preimage: {
                desc: 'Preimage to dry run',
                string: true
            },
            at: {
                desc: 'Block hash to dry run',
                string: true
            },
            'output-path': {
                desc: 'File path to print output',
                string: true
            },
            html: {
                desc: 'Generate html with storage diff'
            },
            open: {
                desc: 'Open generated html'
            }
        }), async (argv)=>{
        if (argv.preimage) {
            await (0, _dryrunpreimage.dryRunPreimage)(argv);
        } else {
            await (0, _dryrunextrinsic.dryRunExtrinsic)(argv);
        }
    });
};
