// @acala-network/chopsticks@0.9.3/dist/cjs/plugins/dry-run/dry-run-preimage.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "dryRunPreimage", {
    enumerable: true,
    get: function() {
        return dryRunPreimage;
    }
});
const _utilcrypto = require("@polkadot/util-crypto");
const _util = require("@polkadot/util");
const _logger = require("../../logger.js");
const _generatehtmldiff = require("../../utils/generate-html-diff.js");
const _chopstickscore = require("@acala-network/chopsticks-core");
const _openhtml = require("../../utils/open-html.js");
const _context = require("../../context.js");
const dryRunPreimage = async (argv)=>{
    const context = await (0, _context.setupContext)(argv);
    const extrinsic = argv['preimage'];
    const block = context.chain.head;
    const registry = await block.registry;
    const header = await (0, _chopstickscore.newHeader)(block);
    const data = (0, _util.hexToU8a)(extrinsic);
    const hash = (0, _utilcrypto.blake2AsHex)(data, 256);
    await (0, _chopstickscore.setStorage)(context.chain, {
        Preimage: {
            PreimageFor: [
                [
                    [
                        [
                            hash,
                            data.byteLength
                        ]
                    ],
                    extrinsic
                ]
            ],
            StatusFor: [
                [
                    [
                        hash
                    ],
                    {
                        Requested: {
                            count: 1,
                            len: data.byteLength
                        }
                    }
                ]
            ]
        },
        Scheduler: {
            Agenda: [
                [
                    [
                        block.number + 1
                    ],
                    [
                        {
                            maybeId: '0x64656d6f637261633a0000000000000000000000000000000000000000000000',
                            priority: 63,
                            call: {
                                Lookup: {
                                    hash: hash,
                                    len: data.byteLength
                                }
                            },
                            origin: {
                                system: {
                                    Root: null
                                }
                            }
                        }
                    ]
                ]
            ],
            Lookup: [
                [
                    [
                        '0x64656d6f637261633a0000000000000000000000000000000000000000000000'
                    ],
                    [
                        block.number + 1,
                        0
                    ]
                ]
            ]
        }
    });
    const calls = [
        [
            'Core_initialize_block',
            [
                header.toHex()
            ]
        ]
    ];
    for (const inherent of (await block.chain.getInherents())){
        calls.push([
            'BlockBuilder_apply_extrinsic',
            [
                inherent
            ]
        ]);
    }
    calls.push([
        'BlockBuilder_finalize_block',
        []
    ]);
    _logger.defaultLogger.info({
        preimage: registry.createType('Call', data).toHuman()
    }, 'Dry run preimage');
    const result = await (0, _chopstickscore.runTask)({
        wasm: await block.wasm,
        calls,
        mockSignatureHost: false,
        allowUnresolvedImports: false,
        runtimeLogLevel: argv['runtime-log-level'] || 0
    }, (0, _chopstickscore.taskHandler)(block));
    if ('Error' in result) {
        throw new Error(result.Error);
    }
    const filePath = await (0, _generatehtmldiff.generateHtmlDiffPreviewFile)(block, result.Call.storageDiff, hash);
    console.log(`Generated preview ${filePath}`);
    if (argv['open']) {
        (0, _openhtml.openHtml)(filePath);
    }
    // if dry-run preimage has extrinsic arguments then dry-run extrinsic
    // this is useful to test something after preimage is applied
    if (argv['extrinsic']) {
        await context.chain.newBlock();
        const input = argv['address'] ? {
            call: argv['extrinsic'],
            address: argv['address']
        } : argv['extrinsic'];
        const { outcome, storageDiff } = await context.chain.dryRunExtrinsic(input);
        if (outcome.isErr) {
            throw new Error(outcome.asErr.toString());
        }
        _logger.defaultLogger.info(outcome.toHuman(), 'dry_run_outcome');
        const filePath = await (0, _generatehtmldiff.generateHtmlDiffPreviewFile)(context.chain.head, storageDiff, (0, _utilcrypto.blake2AsHex)(argv['extrinsic'], 256));
        console.log(`Generated preview ${filePath}`);
        if (argv['open']) {
            (0, _openhtml.openHtml)(filePath);
        }
    }
    process.exit(0);
};
