// @acala-network/chopsticks-core@0.9.3/dist/cjs/rpc/shared.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    ResponseError: function() {
        return ResponseError;
    },
    logger: function() {
        return logger;
    }
});
const _logger = require("../logger.js");
function _define_property(obj, key, value) {
    if (key in obj) {
        Object.defineProperty(obj, key, {
            value: value,
            enumerable: true,
            configurable: true,
            writable: true
        });
    } else {
        obj[key] = value;
    }
    return obj;
}
const logger = _logger.defaultLogger.child({
    name: 'rpc'
});
class ResponseError extends Error {
    toJSON() {
        return {
            code: this.code,
            message: this.message
        };
    }
    constructor(code, message){
        super(message);
        _define_property(this, "code", void 0);
        this.name = 'ResponseError';
        this.code = code;
    }
}
