// @acala-network/chopsticks@0.9.3/dist/cjs/plugins/new-block/rpc.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "rpc", {
    enumerable: true,
    get: function() {
        return rpc;
    }
});
const _chopstickscore = require("@acala-network/chopsticks-core");
const _zod = require("zod");
const _logger = require("../../logger.js");
const _index = require("../../schema/index.js");
const schema = _zod.z.object({
    count: _zod.z.number().optional(),
    to: _zod.z.number().optional(),
    dmp: _zod.z.array(_zod.z.object({
        sentAt: _zod.z.number(),
        msg: _index.zHex
    })).min(1).optional(),
    ump: _zod.z.record(_zod.z.number(), _zod.z.array(_index.zHex).min(1)).optional(),
    hrmp: _zod.z.record(_zod.z.number(), _zod.z.array(_zod.z.object({
        sentAt: _zod.z.number(),
        data: _index.zHex
    })).min(1)).optional(),
    transactions: _zod.z.array(_index.zHex).min(1).optional(),
    unsafeBlockHeight: _zod.z.number().optional()
});
const rpc = async (context, [params])=>{
    const { count, to, hrmp, ump, dmp, transactions, unsafeBlockHeight } = schema.parse(params || {});
    const now = context.chain.head.number;
    const diff = to ? to - now : count;
    const finalCount = diff !== undefined ? Math.max(diff, 1) : 1;
    let finalHash;
    if (unsafeBlockHeight !== undefined && unsafeBlockHeight <= now) {
        throw new _chopstickscore.ResponseError(1, 'unsafeBlockHeight must be greater than current block height');
    }
    for(let i = 0; i < finalCount; i++){
        const block = await context.chain.newBlock({
            transactions,
            horizontalMessages: hrmp,
            upwardMessages: ump,
            downwardMessages: dmp,
            unsafeBlockHeight: i === 0 ? unsafeBlockHeight : undefined
        }).catch((error)=>{
            throw new _chopstickscore.ResponseError(1, error.toString());
        });
        _logger.defaultLogger.debug({
            hash: block.hash
        }, 'dev_newBlock');
        finalHash = block.hash;
    }
    return finalHash;
};
