// @moonwall/cli@4.4.4/dist/chunk-ELWCOTTZ.js
// src/lib/governanceProcedures.ts
import "@moonbeam-network/api-augment";
import { expect } from "vitest";
import { blake2AsHex } from "@polkadot/util-crypto";
import { alith, baltathar, charleth, dorothy } from "@moonwall/util";
var COUNCIL_MEMBERS = [baltathar, charleth, dorothy];
var COUNCIL_THRESHOLD = Math.ceil(COUNCIL_MEMBERS.length * 2 / 3);
var TECHNICAL_COMMITTEE_MEMBERS = [alith, baltathar];
var TECHNICAL_COMMITTEE_THRESHOLD = Math.ceil(
  TECHNICAL_COMMITTEE_MEMBERS.length * 2 / 3
);
var notePreimage = async (context, proposal, account = alith) => {
  const encodedProposal = proposal.method.toHex() || "";
  await context.createBlock(
    context.polkadotJs().tx.preimage.notePreimage(encodedProposal).signAsync(account)
  );
  return blake2AsHex(encodedProposal);
};
var instantFastTrack = async (context, proposal, { votingPeriod, delayPeriod } = { votingPeriod: 2, delayPeriod: 0 }) => {
  const proposalHash = typeof proposal == "string" ? proposal : await notePreimage(context, proposal);
  await execCouncilProposal(
    context,
    context.polkadotJs().tx.democracy.externalProposeMajority({
      Lookup: {
        hash: proposalHash,
        len: typeof proposal == "string" ? proposal : proposal.method.encodedLength
      }
    })
  );
  await execTechnicalCommitteeProposal(
    context,
    context.polkadotJs().tx.democracy.fastTrack(proposalHash, votingPeriod, delayPeriod)
  );
  return proposalHash;
};
var execCouncilProposal = async (context, polkadotCall, index = -1, voters = COUNCIL_MEMBERS, threshold = COUNCIL_THRESHOLD) => {
  const lengthBound = polkadotCall.method.encodedLength;
  const { result: proposalResult } = await context.createBlock(
    context.polkadotJs().tx.councilCollective.propose(threshold, polkadotCall, lengthBound).signAsync(charleth)
  );
  if (threshold <= 1) {
    return proposalResult;
  }
  expect(proposalResult.successful, `Council proposal refused: ${proposalResult?.error?.name}`).to.be.true;
  const proposalHash = proposalResult.events.find(({ event: { method } }) => method.toString() == "Proposed").event.data[2].toHex();
  const proposalIndex = index >= 0 ? index : (await context.polkadotJs().query.councilCollective.proposalCount()).toNumber() - 1;
  await Promise.all(
    voters.map(
      (voter) => context.polkadotJs().tx.councilCollective.vote(proposalHash, proposalIndex, true).signAndSend(voter)
    )
  );
  await context.createBlock();
  return await context.createBlock(
    context.polkadotJs().tx.councilCollective.close(
      proposalHash,
      proposalIndex,
      {
        refTime: 2e9,
        proofSize: 1e5
      },
      lengthBound
    ).signAsync(dorothy)
  );
};
var proposeReferendaAndDeposit = async (context, decisionDepositer, proposal, origin) => {
  const proposalHash = typeof proposal == "string" ? proposal : await notePreimage(context, proposal);
  const { result: proposalResult } = await context.createBlock(
    context.polkadotJs().tx.referenda.submit(
      origin,
      {
        Lookup: {
          hash: proposalHash,
          len: typeof proposal == "string" ? proposal : proposal.method.encodedLength
        }
      },
      { At: 0 }
    ).signAsync(alith)
  );
  expect(proposalResult.successful, `Unable to post referenda: ${proposalResult?.error?.name}`).to.be.true;
  const refIndex = proposalResult.events.find(({ event: { method } }) => method.toString() == "Submitted").event.data[0].toString();
  await context.createBlock(
    context.polkadotJs().tx.referenda.placeDecisionDeposit(refIndex).signAsync(decisionDepositer)
  );
  return [+refIndex, proposalHash];
};
var dispatchAsGeneralAdmin = async (context, call) => {
  await context.createBlock(
    context.polkadotJs().tx.sudo.sudo(
      context.polkadotJs().tx.utility.dispatchAs(
        {
          Origins: "GeneralAdmin"
        },
        call
      )
    )
  );
};
var maximizeConvictionVotingOf = async (context, voters, refIndex) => {
  const fee = (await context.polkadotJs().tx.convictionVoting.vote(refIndex, {
    Standard: {
      vote: { aye: true, conviction: "Locked6x" },
      balance: (await context.polkadotJs().query.system.account(alith.address)).data.free
    }
  }).paymentInfo(alith)).partialFee;
  await context.createBlock(
    voters.map(
      async (voter) => context.polkadotJs().tx.convictionVoting.vote(refIndex, {
        Standard: {
          vote: { aye: true, conviction: "Locked6x" },
          balance: await (await context.polkadotJs().query.system.account(voter.address)).data.free.sub(fee)
        }
      }).signAsync(voter)
    )
  );
};
var execTechnicalCommitteeProposal = async (context, polkadotCall, voters = TECHNICAL_COMMITTEE_MEMBERS, threshold = TECHNICAL_COMMITTEE_THRESHOLD) => {
  const lengthBound = polkadotCall.encodedLength;
  const { result: proposalResult } = await context.createBlock(
    context.polkadotJs().tx.techCommitteeCollective.propose(threshold, polkadotCall, lengthBound)
  );
  if (threshold <= 1) {
    return proposalResult;
  }
  expect(proposalResult.successful, `Council proposal refused: ${proposalResult?.error?.name}`).to.be.true;
  const proposalHash = proposalResult.events.find(({ event: { method } }) => method.toString() == "Proposed").event.data[2].toHex();
  const proposalCount = await context.polkadotJs().query.techCommitteeCollective.proposalCount();
  await context.createBlock(
    voters.map(
      (voter) => context.polkadotJs().tx.techCommitteeCollective.vote(proposalHash, Number(proposalCount) - 1, true).signAsync(voter)
    )
  );
  const { result: closeResult } = await context.createBlock(
    context.polkadotJs().tx.techCommitteeCollective.close(
      proposalHash,
      Number(proposalCount) - 1,
      {
        refTime: 2e9,
        proofSize: 1e5
      },
      lengthBound
    ).signAsync(baltathar)
  );
  return closeResult;
};
var executeProposalWithCouncil = async (api, encodedHash) => {
  let nonce = (await api.rpc.system.accountNextIndex(alith.address)).toNumber();
  const referendumNextIndex = (await api.query.democracy.referendumCount()).toNumber();
  const callData = api.consts.system.version.specVersion.toNumber() >= 2e3 ? { Legacy: encodedHash } : encodedHash;
  const external = api.tx.democracy.externalProposeMajority(callData);
  const fastTrack = api.tx.democracy.fastTrack(encodedHash, 1, 0);
  const voteAmount = 1n * 10n ** BigInt(api.registry.chainDecimals[0]);
  process.stdout.write(`Sending motion + fast-track + vote for ${encodedHash}...`);
  await Promise.all([
    api.tx.councilCollective.propose(1, external, external.length).signAndSend(alith, { nonce: nonce++ }),
    api.tx.techCommitteeCollective.propose(1, fastTrack, fastTrack.length).signAndSend(alith, { nonce: nonce++ }),
    api.tx.democracy.vote(referendumNextIndex, {
      Standard: {
        balance: voteAmount,
        vote: { aye: true, conviction: 1 }
      }
    }).signAndSend(alith, { nonce: nonce++ })
  ]);
  process.stdout.write(`\u2705
`);
  process.stdout.write(`Waiting for referendum [${referendumNextIndex}] to be executed...`);
  let referenda;
  while (!referenda) {
    referenda = (await api.query.democracy.referendumInfoOf.entries()).find(
      (ref) => ref[1].unwrap().isFinished && api.registry.createType("u32", ref[0].toU8a().slice(-4)).toNumber() == referendumNextIndex
    )?.[1].unwrap();
    await new Promise((resolve) => setTimeout(resolve, 1e3));
  }
  process.stdout.write(`${referenda.asFinished.approved ? `\u2705` : `\u274C`} 
`);
  if (!referenda.asFinished.approved) {
    throw new Error("Finished Referendum was not approved");
  }
};
var cancelReferendaWithCouncil = async (api, refIndex) => {
  const proposal = api.tx.democracy.cancelReferendum(refIndex);
  const encodedProposal = proposal.method.toHex();
  const encodedHash = blake2AsHex(encodedProposal);
  let nonce = (await api.rpc.system.accountNextIndex(alith.address)).toNumber();
  await api.tx.democracy.notePreimage(encodedProposal).signAndSend(alith, { nonce: nonce++ });
  await executeProposalWithCouncil(api, encodedHash);
};

export {
  COUNCIL_MEMBERS,
  COUNCIL_THRESHOLD,
  TECHNICAL_COMMITTEE_MEMBERS,
  TECHNICAL_COMMITTEE_THRESHOLD,
  notePreimage,
  instantFastTrack,
  execCouncilProposal,
  proposeReferendaAndDeposit,
  dispatchAsGeneralAdmin,
  maximizeConvictionVotingOf,
  execTechnicalCommitteeProposal,
  executeProposalWithCouncil,
  cancelReferendaWithCouncil
};
