// @acala-network/chopsticks-core@0.9.3/dist/esm/utils/decoder.js
import '@polkadot/types-codec';
import { LRUCache } from 'lru-cache';
import { hexToU8a, u8aToHex } from '@polkadot/util';
import _ from 'lodash';
import { decodeWellKnownKey } from './well-known-keys.js';
const _CACHE = {};
function createCache() {
    return new LRUCache({
        max: 50
    });
}
const getCache = (uid)=>{
    if (!_CACHE[uid]) {
        _CACHE[uid] = createCache();
    }
    return _CACHE[uid];
};
const getStorageEntry = (meta, block, key)=>{
    const cache = getCache(block.chain.uid);
    for (const prefix of cache.keys()){
        if (key.startsWith(prefix)) // update the recency of the cache entry
        return cache.get(prefix);
    }
    for (const module of Object.values(meta.query)){
        for (const storage of Object.values(module)){
            const keyPrefix = u8aToHex(storage.keyPrefix());
            if (key.startsWith(keyPrefix)) {
                cache.set(keyPrefix, storage);
                return storage;
            }
        }
    }
    return undefined;
};
export const decodeKey = (meta, block, key)=>{
    const storage = getStorageEntry(meta, block, key);
    const decodedKey = meta.registry.createType('StorageKey', key);
    if (storage) {
        decodedKey.setMeta(storage.meta);
        return {
            storage,
            decodedKey
        };
    }
    return {};
};
export const decodeKeyValue = (meta, block, key, value, toHuman = true)=>{
    const res = decodeWellKnownKey(meta.registry, key, value);
    if (res) {
        return {
            section: 'substrate',
            method: res.name,
            key: res.key,
            value: res.value
        };
    }
    const { storage, decodedKey } = decodeKey(meta, block, key);
    if (!storage || !decodedKey) {
        return undefined;
    }
    const decodeValue = ()=>{
        if (!value) return null;
        return meta.registry.createType(decodedKey.outputType, hexToU8a(value))[toHuman ? 'toHuman' : 'toJSON']();
    };
    return {
        section: storage.section,
        method: storage.method,
        key: decodedKey.args.map((x)=>x.toJSON()),
        value: decodeValue()
    };
};
export const toStorageObject = (decoded)=>{
    if (!decoded) {
        return undefined;
    }
    const { section, method, key, value } = decoded;
    let obj = value;
    if (key) {
        for(let i = key.length - 1; i >= 0; i--){
            const k = key[i];
            const strKey = [
                'string',
                'number'
            ].includes(typeof k) ? k : JSON.stringify(k);
            const newObj = {
                [strKey]: obj
            };
            obj = newObj;
        }
    }
    return {
        [section]: {
            [method]: obj
        }
    };
};
/**
 * Decode block storage diff
 * @param block Block to compare storage diff
 * @param diff Storage diff
 * @returns decoded old state and new state
 */ export const decodeBlockStorageDiff = async (block, diff)=>{
    const oldState = {};
    const newState = {};
    const meta = await block.meta;
    for (const [key, value] of diff){
        const oldValue = await block.get(key);
        const oldDecoded = toStorageObject(decodeKeyValue(meta, block, key, oldValue)) ?? {
            [key]: oldValue
        };
        _.merge(oldState, oldDecoded);
        const newDecoded = toStorageObject(decodeKeyValue(meta, block, key, value)) ?? {
            [key]: value
        };
        _.merge(newState, newDecoded);
    }
    return [
        oldState,
        newState
    ];
};
