// @polkadot/types@10.11.1/cjs/metadata/decorate/storage/util.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createRuntimeFunction = void 0;
const createFunction_js_1 = require("./createFunction.js");
function findSiPrimitive(registry, _prim) {
    const prim = _prim.toLowerCase();
    return registry.lookup.types.find((t) => (t.type.def.isPrimitive &&
        t.type.def.asPrimitive.toString().toLowerCase() === prim) || (t.type.def.isHistoricMetaCompat &&
        t.type.def.asHistoricMetaCompat.toString().toLowerCase() === prim));
}
function findSiType(registry, orig) {
    let portable = findSiPrimitive(registry, orig);
    if (!portable && orig === 'Bytes') {
        const u8 = findSiPrimitive(registry, 'u8');
        if (u8) {
            portable = registry.lookup.types.find((t) => (t.type.def.isSequence &&
                t.type.def.asSequence.type.eq(u8.id)) || (t.type.def.isHistoricMetaCompat &&
                t.type.def.asHistoricMetaCompat.eq(orig)));
        }
    }
    if (!portable) {
        console.warn(`Unable to map ${orig} to a lookup index`);
    }
    return portable;
}
/** @internal */
function createRuntimeFunction({ method, prefix, section }, key, { docs, type }) {
    return (registry) => (0, createFunction_js_1.createFunction)(registry, {
        meta: registry.createTypeUnsafe('StorageEntryMetadataLatest', [{
                docs: registry.createTypeUnsafe('Vec<Text>', [[docs]]),
                modifier: registry.createTypeUnsafe('StorageEntryModifierLatest', ['Required']),
                name: registry.createTypeUnsafe('Text', [method]),
                toJSON: () => key,
                type: registry.createTypeUnsafe('StorageEntryTypeLatest', [{ Plain: findSiType(registry, type)?.id || 0 }])
            }]),
        method,
        prefix,
        section
    }, { key, skipHashing: true });
}
exports.createRuntimeFunction = createRuntimeFunction;
