// @acala-network/chopsticks-core@0.9.3/dist/cjs/blockchain/block-builder.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    buildBlock: function() {
        return buildBlock;
    },
    dryRunExtrinsic: function() {
        return dryRunExtrinsic;
    },
    dryRunInherents: function() {
        return dryRunInherents;
    },
    newHeader: function() {
        return newHeader;
    }
});
const _block = require("./block.js");
const _storagelayer = require("./storage-layer.js");
const _util = require("@polkadot/util");
const _index = require("../utils/index.js");
const _logger = require("../logger.js");
const _timetravel = require("../utils/time-travel.js");
const logger = _logger.defaultLogger.child({
    name: 'block-builder'
});
const getConsensus = (header)=>{
    if (header.digest.logs.length === 0) return;
    const preRuntime = header.digest.logs[0].asPreRuntime;
    const [consensusEngine, slot] = preRuntime;
    return {
        consensusEngine,
        slot,
        rest: header.digest.logs.slice(1)
    };
};
const getNewSlot = (digest, slotNumber)=>{
    if (digest.isPrimary) {
        return {
            primary: {
                ...digest.asPrimary.toJSON(),
                slotNumber
            }
        };
    }
    if (digest.isSecondaryPlain) {
        return {
            secondaryPlain: {
                ...digest.asSecondaryPlain.toJSON(),
                slotNumber
            }
        };
    }
    if (digest.isSecondaryVRF) {
        return {
            secondaryVRF: {
                ...digest.asSecondaryVRF.toJSON(),
                slotNumber
            }
        };
    }
    return digest.toJSON();
};
const newHeader = async (head, unsafeBlockHeight)=>{
    const meta = await head.meta;
    const parentHeader = await head.header;
    let newLogs = parentHeader.digest.logs;
    const consensus = getConsensus(parentHeader);
    if (consensus?.consensusEngine.isAura) {
        const slot = await (0, _timetravel.getCurrentSlot)(head.chain);
        const newSlot = (0, _util.compactAddLength)(meta.registry.createType('Slot', slot + 1).toU8a());
        newLogs = [
            {
                PreRuntime: [
                    consensus.consensusEngine,
                    newSlot
                ]
            },
            ...consensus.rest
        ];
    } else if (consensus?.consensusEngine.isBabe) {
        const slot = await (0, _timetravel.getCurrentSlot)(head.chain);
        const digest = meta.registry.createType('RawBabePreDigest', consensus.slot);
        const newSlot = (0, _util.compactAddLength)(meta.registry.createType('RawBabePreDigest', getNewSlot(digest, slot + 1)).toU8a());
        newLogs = [
            {
                PreRuntime: [
                    consensus.consensusEngine,
                    newSlot
                ]
            },
            ...consensus.rest
        ];
    } else if (consensus?.consensusEngine?.toString() == 'nmbs') {
        const nmbsKey = (0, _util.stringToHex)('nmbs');
        newLogs = [
            {
                // Using previous block author
                PreRuntime: [
                    consensus.consensusEngine,
                    parentHeader.digest.logs.find((log)=>log.isPreRuntime && log.asPreRuntime[0].toHex() == nmbsKey)?.asPreRuntime[1].toHex()
                ]
            },
            ...consensus.rest
        ];
        if (meta.query.randomness?.notFirstBlock) {
            // TODO: shouldn't modify existing head
            // reset notFirstBlock so randomness will skip validation
            head.pushStorageLayer().set((0, _index.compactHex)(meta.query.randomness.notFirstBlock()), _storagelayer.StorageValueKind.Deleted);
        }
    }
    const header = meta.registry.createType('Header', {
        parentHash: head.hash,
        number: unsafeBlockHeight ?? head.number + 1,
        stateRoot: '0x0000000000000000000000000000000000000000000000000000000000000000',
        extrinsicsRoot: '0x0000000000000000000000000000000000000000000000000000000000000000',
        digest: {
            logs: newLogs
        }
    });
    return header;
};
const initNewBlock = async (head, header, inherents, storageLayer, callback)=>{
    const blockNumber = header.number.toNumber();
    const hash = `0x${Math.round(Math.random() * 100000000).toString(16).padEnd(64, '0')}`;
    const newBlock = new _block.Block(head.chain, blockNumber, hash, head, {
        header,
        extrinsics: [],
        storage: storageLayer ?? head.storage
    });
    {
        // initialize block
        const resp = await newBlock.call('Core_initialize_block', [
            header.toHex()
        ]);
        newBlock.pushStorageLayer().setAll(resp.storageDiff);
        callback?.onPhaseApplied?.('initialize', resp);
    }
    const layers = [];
    // apply inherents
    for (const extrinsic of inherents){
        try {
            const resp = await newBlock.call('BlockBuilder_apply_extrinsic', [
                extrinsic
            ]);
            const layer = newBlock.pushStorageLayer();
            layer.setAll(resp.storageDiff);
            layers.push(layer);
            callback?.onPhaseApplied?.(layers.length - 1, resp);
        } catch (e) {
            logger.warn('Failed to apply inherents %o %s', e, e);
            throw new Error('Failed to apply inherents');
        }
    }
    return {
        block: newBlock,
        layers: layers
    };
};
const buildBlock = async (head, inherents, extrinsics, ump, callbacks, unsafeBlockHeight)=>{
    const registry = await head.registry;
    const header = await newHeader(head, unsafeBlockHeight);
    const newBlockNumber = header.number.toNumber();
    logger.info({
        number: newBlockNumber,
        extrinsicsCount: extrinsics.length,
        umpCount: Object.keys(ump).length
    }, `Try building block #${newBlockNumber.toLocaleString()}`);
    let layer;
    // apply ump via storage override hack
    if (Object.keys(ump).length > 0) {
        const meta = await head.meta;
        layer = new _storagelayer.StorageLayer(head.storage);
        for (const [paraId, upwardMessages] of Object.entries(ump)){
            const upwardMessagesU8a = upwardMessages.map((x)=>(0, _util.hexToU8a)(x));
            const messagesCount = upwardMessages.length;
            const messagesSize = upwardMessagesU8a.map((x)=>x.length).reduce((s, i)=>s + i, 0);
            if (meta.query.ump) {
                const queueSize = meta.registry.createType('(u32, u32)', [
                    messagesCount,
                    messagesSize
                ]);
                const messages = meta.registry.createType('Vec<Bytes>', upwardMessages);
                // TODO: make sure we append instead of replace
                layer.setAll([
                    [
                        (0, _index.compactHex)(meta.query.ump.relayDispatchQueues(paraId)),
                        messages.toHex()
                    ],
                    [
                        (0, _index.compactHex)(meta.query.ump.relayDispatchQueueSize(paraId)),
                        queueSize.toHex()
                    ]
                ]);
            } else if (meta.query.messageQueue) {
                // TODO: make sure we append instead of replace
                const origin = {
                    ump: {
                        para: paraId
                    }
                };
                let last = 0;
                let heap = new Uint8Array(0);
                for (const message of upwardMessagesU8a){
                    const payloadLen = message.length;
                    const header = meta.registry.createType('(u32, bool)', [
                        payloadLen,
                        false
                    ]);
                    last = heap.length;
                    heap = (0, _util.u8aConcat)(heap, header.toU8a(), message);
                }
                layer.setAll([
                    [
                        (0, _index.compactHex)(meta.query.messageQueue.bookStateFor(origin)),
                        meta.registry.createType('PalletMessageQueueBookState', {
                            begin: 0,
                            end: 1,
                            count: 1,
                            readyNeighbours: {
                                prev: origin,
                                next: origin
                            },
                            messageCount: messagesCount,
                            size_: messagesSize
                        }).toHex()
                    ],
                    [
                        (0, _index.compactHex)(meta.query.messageQueue.serviceHead(origin)),
                        meta.registry.createType('PolkadotRuntimeParachainsInclusionAggregateMessageOrigin', origin).toHex()
                    ],
                    [
                        (0, _index.compactHex)(meta.query.messageQueue.pages(origin, 0)),
                        meta.registry.createType('PalletMessageQueuePage', {
                            remaining: messagesCount,
                            remaining_size: messagesSize,
                            first_index: 0,
                            first: 0,
                            last,
                            heap: (0, _util.compactAddLength)(heap)
                        }).toHex()
                    ]
                ]);
            } else {
                throw new Error('Unknown ump storage');
            }
            logger.trace({
                paraId,
                upwardMessages: (0, _logger.truncate)(upwardMessages)
            }, 'Pushed UMP');
        }
        if (meta.query.ump) {
            const needsDispatch = meta.registry.createType('Vec<u32>', Object.keys(ump));
            layer.set((0, _index.compactHex)(meta.query.ump.needsDispatch()), needsDispatch.toHex());
        }
    }
    const { block: newBlock } = await initNewBlock(head, header, inherents, layer);
    const pendingExtrinsics = [];
    const includedExtrinsic = [];
    // apply extrinsics
    for (const extrinsic of extrinsics){
        try {
            const resp = await newBlock.call('BlockBuilder_apply_extrinsic', [
                extrinsic
            ]);
            const outcome = registry.createType('ApplyExtrinsicResult', resp.result);
            if (outcome.isErr) {
                callbacks?.onApplyExtrinsicError?.(extrinsic, outcome.asErr);
                continue;
            }
            newBlock.pushStorageLayer().setAll(resp.storageDiff);
            includedExtrinsic.push(extrinsic);
            callbacks?.onPhaseApplied?.(includedExtrinsic.length - 1, resp);
        } catch (e) {
            logger.info('Failed to apply extrinsic %o %s', e, e);
            pendingExtrinsics.push(extrinsic);
        }
    }
    {
        // finalize block
        const resp = await newBlock.call('BlockBuilder_finalize_block', []);
        newBlock.pushStorageLayer().setAll(resp.storageDiff);
        callbacks?.onPhaseApplied?.('finalize', resp);
    }
    const blockData = registry.createType('Block', {
        header,
        extrinsics: includedExtrinsic
    });
    const storageDiff = await newBlock.storageDiff();
    if (logger.level.toLowerCase() === 'trace') {
        logger.trace(Object.entries(storageDiff).map(([key, value])=>[
                key,
                (0, _logger.truncate)(value)
            ]), 'Final block');
    }
    const finalBlock = new _block.Block(head.chain, newBlock.number, blockData.hash.toHex(), head, {
        header,
        extrinsics: [
            ...inherents,
            ...includedExtrinsic
        ],
        storage: head.storage,
        storageDiff
    });
    logger.info({
        number: newBlock.number,
        hash: finalBlock.hash,
        extrinsics: (0, _logger.truncate)(includedExtrinsic),
        pendingExtrinsicsCount: pendingExtrinsics.length,
        ump: (0, _logger.truncate)(ump)
    }, 'Block built');
    return [
        finalBlock,
        pendingExtrinsics
    ];
};
const dryRunExtrinsic = async (head, inherents, extrinsic)=>{
    const registry = await head.registry;
    const header = await newHeader(head);
    const { block: newBlock } = await initNewBlock(head, header, inherents);
    if (typeof extrinsic !== 'string') {
        if (!head.chain.mockSignatureHost) {
            throw new Error('Cannot fake signature because mock signature host is not enabled. Start chain with `mockSignatureHost: true`');
        }
        const meta = await head.meta;
        const call = registry.createType('Call', (0, _util.hexToU8a)(extrinsic.call));
        const generic = registry.createType('GenericExtrinsic', call);
        const accountRaw = await head.get((0, _index.compactHex)(meta.query.system.account(extrinsic.address)));
        const account = registry.createType('AccountInfo', (0, _util.hexToU8a)(accountRaw));
        generic.signFake(extrinsic.address, {
            blockHash: head.hash,
            genesisHash: head.hash,
            runtimeVersion: await head.runtimeVersion,
            nonce: account.nonce
        });
        const mockSignature = new Uint8Array(64);
        mockSignature.fill(0xcd);
        mockSignature.set([
            0xde,
            0xad,
            0xbe,
            0xef
        ]);
        generic.signature.set(mockSignature);
        logger.debug({
            call: call.toHuman()
        }, 'dry_run_call');
        return newBlock.call('BlockBuilder_apply_extrinsic', [
            generic.toHex()
        ]);
    }
    logger.debug({
        call: registry.createType('GenericExtrinsic', (0, _util.hexToU8a)(extrinsic)).toHuman()
    }, 'dry_run_extrinsic');
    return newBlock.call('BlockBuilder_apply_extrinsic', [
        extrinsic
    ]);
};
const dryRunInherents = async (head, inherents)=>{
    const header = await newHeader(head);
    const { layers } = await initNewBlock(head, header, inherents);
    const storage = {};
    for (const layer of layers){
        await layer.mergeInto(storage);
    }
    return Object.entries(storage);
};
