// @acala-network/chopsticks-core@0.9.3/dist/cjs/api.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "Api", {
    enumerable: true,
    get: function() {
        return Api;
    }
});
const _index = require("./utils/index.js");
function _check_private_redeclaration(obj, privateCollection) {
    if (privateCollection.has(obj)) {
        throw new TypeError("Cannot initialize the same private elements twice on an object");
    }
}
function _class_apply_descriptor_get(receiver, descriptor) {
    if (descriptor.get) {
        return descriptor.get.call(receiver);
    }
    return descriptor.value;
}
function _class_apply_descriptor_set(receiver, descriptor, value) {
    if (descriptor.set) {
        descriptor.set.call(receiver, value);
    } else {
        if (!descriptor.writable) {
            throw new TypeError("attempted to set read only private field");
        }
        descriptor.value = value;
    }
}
function _class_extract_field_descriptor(receiver, privateMap, action) {
    if (!privateMap.has(receiver)) {
        throw new TypeError("attempted to " + action + " private field on non-instance");
    }
    return privateMap.get(receiver);
}
function _class_private_field_get(receiver, privateMap) {
    var descriptor = _class_extract_field_descriptor(receiver, privateMap, "get");
    return _class_apply_descriptor_get(receiver, descriptor);
}
function _class_private_field_init(obj, privateMap, value) {
    _check_private_redeclaration(obj, privateMap);
    privateMap.set(obj, value);
}
function _class_private_field_set(receiver, privateMap, value) {
    var descriptor = _class_extract_field_descriptor(receiver, privateMap, "set");
    _class_apply_descriptor_set(receiver, descriptor, value);
    return value;
}
function _define_property(obj, key, value) {
    if (key in obj) {
        Object.defineProperty(obj, key, {
            value: value,
            enumerable: true,
            configurable: true,
            writable: true
        });
    } else {
        obj[key] = value;
    }
    return obj;
}
var _provider = /*#__PURE__*/ new WeakMap(), _ready = /*#__PURE__*/ new WeakMap(), _chain = /*#__PURE__*/ new WeakMap(), _chainProperties = /*#__PURE__*/ new WeakMap();
class Api {
    async disconnect() {
        return _class_private_field_get(this, _provider).disconnect();
    }
    get isReady() {
        if (!_class_private_field_get(this, _ready)) {
            if (_class_private_field_get(this, _provider)['isReady']) {
                _class_private_field_set(this, _ready, _class_private_field_get(this, _provider)['isReady']);
            } else {
                _class_private_field_set(this, _ready, new Promise((resolve)=>{
                    if (_class_private_field_get(this, _provider).hasSubscriptions) {
                        _class_private_field_get(this, _provider).on('connected', resolve);
                        _class_private_field_get(this, _provider).connect();
                    } else {
                        resolve();
                    }
                }));
            }
        }
        return _class_private_field_get(this, _ready);
    }
    get chain() {
        if (!_class_private_field_get(this, _chain)) {
            _class_private_field_set(this, _chain, this.getSystemChain());
        }
        return _class_private_field_get(this, _chain);
    }
    get chainProperties() {
        if (!_class_private_field_get(this, _chainProperties)) {
            _class_private_field_set(this, _chainProperties, this.getSystemProperties());
        }
        return _class_private_field_get(this, _chainProperties);
    }
    async getSystemName() {
        return _class_private_field_get(this, _provider).send('system_name', []);
    }
    async getSystemProperties() {
        return _class_private_field_get(this, _provider).send('system_properties', []);
    }
    async getSystemChain() {
        return _class_private_field_get(this, _provider).send('system_chain', []);
    }
    async getBlockHash(blockNumber) {
        return _class_private_field_get(this, _provider).send('chain_getBlockHash', Number.isInteger(blockNumber) ? [
            blockNumber
        ] : [], !!blockNumber);
    }
    async getHeader(hash) {
        return _class_private_field_get(this, _provider).send('chain_getHeader', hash ? [
            hash
        ] : [], !!hash);
    }
    async getBlock(hash) {
        return _class_private_field_get(this, _provider).send('chain_getBlock', hash ? [
            hash
        ] : [], !!hash);
    }
    async getStorage(key, hash) {
        const [child, storageKey] = (0, _index.splitChildKey)(key);
        if (child) {
            // child storage key, use childstate_getStorage
            const params = [
                child,
                storageKey
            ];
            if (hash) params.push(hash);
            return _class_private_field_get(this, _provider).send('childstate_getStorage', params, !!hash);
        } else {
            // main storage key, use state_getStorage
            const params = [
                key
            ];
            if (hash) params.push(hash);
            return _class_private_field_get(this, _provider).send('state_getStorage', params, !!hash);
        }
    }
    async getKeysPaged(prefix, pageSize, startKey, hash) {
        const [child, storageKey] = (0, _index.splitChildKey)(prefix);
        if (child) {
            // child storage key, use childstate_getKeysPaged
            // strip child prefix from startKey
            const params = [
                child,
                storageKey,
                pageSize,
                (0, _index.stripChildPrefix)(startKey)
            ];
            if (hash) params.push(hash);
            return _class_private_field_get(this, _provider).send('childstate_getKeysPaged', params, !!hash).then((keys)=>keys.map((key)=>(0, _index.prefixedChildKey)(child, key)));
        } else {
            // main storage key, use state_getKeysPaged
            const params = [
                prefix,
                pageSize,
                startKey
            ];
            if (hash) params.push(hash);
            return _class_private_field_get(this, _provider).send('state_getKeysPaged', params, !!hash);
        }
    }
    async subscribeRemoteNewHeads(cb) {
        if (!_class_private_field_get(this, _provider).hasSubscriptions) {
            throw new Error('subscribeRemoteNewHeads only works with subscriptions');
        }
        return _class_private_field_get(this, _provider).subscribe('chain_newHead', 'chain_subscribeNewHeads', [], cb);
    }
    async subscribeRemoteFinalizedHeads(cb) {
        if (!_class_private_field_get(this, _provider).hasSubscriptions) {
            throw new Error('subscribeRemoteFinalizedHeads only works with subscriptions');
        }
        return _class_private_field_get(this, _provider).subscribe('chain_finalizedHead', 'chain_subscribeFinalizedHeads', [], cb);
    }
    constructor(provider, signedExtensions){
        _class_private_field_init(this, _provider, {
            writable: true,
            value: void 0
        });
        _class_private_field_init(this, _ready, {
            writable: true,
            value: void 0
        });
        _class_private_field_init(this, _chain, {
            writable: true,
            value: void 0
        });
        _class_private_field_init(this, _chainProperties, {
            writable: true,
            value: void 0
        });
        _define_property(this, "signedExtensions", void 0);
        _class_private_field_set(this, _provider, provider);
        this.signedExtensions = signedExtensions || {};
    }
}
