// @acala-network/chopsticks@0.9.3/dist/cjs/plugins/types.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    dryRun: function() {
        return _index1.rpc;
    },
    newBlock: function() {
        return _index.rpc;
    },
    runBlock: function() {
        return _index7.rpc;
    },
    setBlockBuildMode: function() {
        return _index2.rpc;
    },
    setHead: function() {
        return _index3.rpc;
    },
    setRuntimeLogLevel: function() {
        return _index4.rpc;
    },
    setStorage: function() {
        return _index5.rpc;
    },
    timeTravel: function() {
        return _index6.rpc;
    }
});
const _index = require("./new-block/index.js");
const _index1 = require("./dry-run/index.js");
const _index2 = require("./set-block-build-mode/index.js");
const _index3 = require("./set-head/index.js");
const _index4 = require("./set-runtime-log-level/index.js");
const _index5 = require("./set-storage/index.js");
const _index6 = require("./time-travel/index.js");
const _index7 = require("./run-block/index.js");
