// @acala-network/chopsticks-core@0.9.3/dist/cjs/wasm-executor/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    calculateStateRoot: function() {
        return calculateStateRoot;
    },
    createProof: function() {
        return createProof;
    },
    decodeProof: function() {
        return decodeProof;
    },
    destroyWorker: function() {
        return destroyWorker;
    },
    emptyTaskHandler: function() {
        return emptyTaskHandler;
    },
    getAuraSlotDuration: function() {
        return getAuraSlotDuration;
    },
    getRuntimeVersion: function() {
        return getRuntimeVersion;
    },
    runTask: function() {
        return runTask;
    },
    taskHandler: function() {
        return taskHandler;
    }
});
const _comlink = /*#__PURE__*/ _interop_require_wildcard(require("comlink"));
const _util = require("@polkadot/util");
const _utilcrypto = require("@polkadot/util-crypto");
const _lodash = /*#__PURE__*/ _interop_require_default(require("lodash"));
const _keycache = require("../utils/key-cache.js");
const _logger = require("../logger.js");
const _index = require("../utils/index.js");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const logger = _logger.defaultLogger.child({
    name: 'executor'
});
let __executor_worker;
const getWorker = async ()=>{
    if (__executor_worker) return __executor_worker;
    const isNode = typeof process !== 'undefined' && process?.versions?.node // true for node or bun
    ;
    if (isNode) {
        __executor_worker = Promise.resolve().then(()=>/*#__PURE__*/ _interop_require_wildcard(require("./node-worker.js"))).then(({ startWorker })=>startWorker());
    } else {
        __executor_worker = Promise.resolve().then(()=>/*#__PURE__*/ _interop_require_wildcard(require("./browser-worker.js"))).then(({ startWorker })=>startWorker());
    }
    return __executor_worker;
};
const getRuntimeVersion = _lodash.default.memoize(async (code)=>{
    const worker = await getWorker();
    return worker.remote.getRuntimeVersion(code).then((version)=>{
        version.specName = (0, _util.hexToString)(version.specName);
        version.implName = (0, _util.hexToString)(version.implName);
        return version;
    });
});
const calculateStateRoot = async (entries, trie_version)=>{
    const worker = await getWorker();
    return worker.remote.calculateStateRoot(entries, trie_version);
};
const decodeProof = async (trieRootHash, keys, nodes)=>{
    const worker = await getWorker();
    const result = await worker.remote.decodeProof(trieRootHash, keys, nodes);
    return result.reduce((accum, [key, value])=>{
        accum[key] = value;
        return accum;
    }, {});
};
const createProof = async (nodes, entries)=>{
    const worker = await getWorker();
    const [trieRootHash, newNodes] = await worker.remote.createProof(nodes, entries);
    return {
        trieRootHash,
        nodes: newNodes
    };
};
const runTask = async (task, callback = emptyTaskHandler)=>{
    const worker = await getWorker();
    logger.trace((0, _logger.truncate)(task), 'taskRun');
    const response = await worker.remote.runTask(task, _comlink.proxy(callback));
    if ('Call' in response) {
        logger.trace((0, _logger.truncate)(response.Call), 'taskResponse');
    } else {
        logger.trace({
            response
        }, 'taskResponse');
    }
    return response;
};
const taskHandler = (block)=>{
    return {
        getStorage: async function(key) {
            return block.get(key);
        },
        getStateRoot: async function() {
            const header = await block.header;
            return header.stateRoot.toHex();
        },
        getNextKey: async function(prefix, key) {
            const [nextKey] = await block.getKeysPaged({
                prefix: prefix.length === 2 /** 0x */  ? key.slice(0, _keycache.PREFIX_LENGTH) : prefix,
                pageSize: 1,
                startKey: key
            });
            return nextKey && (0, _index.stripChildPrefix)(nextKey);
        },
        offchainGetStorage: async function(key) {
            if (!block.chain.offchainWorker) throw new Error('offchain worker not found');
            return block.chain.offchainWorker.get(key);
        },
        offchainTimestamp: async function() {
            return Date.now();
        },
        offchainRandomSeed: async function() {
            return (0, _utilcrypto.randomAsHex)(32);
        },
        offchainSubmitTransaction: async function(tx) {
            if (!block.chain.offchainWorker) throw new Error('offchain worker not found');
            try {
                const hash = await block.chain.offchainWorker.pushExtrinsic(block, tx);
                logger.trace({
                    hash
                }, 'offchainSubmitTransaction');
                return true;
            } catch (error) {
                logger.trace({
                    error
                }, 'offchainSubmitTransaction');
                return false;
            }
        }
    };
};
const emptyTaskHandler = {
    getStorage: async function(_key) {
        throw new Error('Method not implemented');
    },
    getStateRoot: async function() {
        throw new Error('Method not implemented');
    },
    getNextKey: async function(_prefix, _key) {
        throw new Error('Method not implemented');
    },
    offchainGetStorage: async function(_key) {
        throw new Error('Method not implemented');
    },
    offchainTimestamp: async function() {
        throw new Error('Method not implemented');
    },
    offchainRandomSeed: async function() {
        throw new Error('Method not implemented');
    },
    offchainSubmitTransaction: async function(_tx) {
        throw new Error('Method not implemented');
    }
};
const getAuraSlotDuration = _lodash.default.memoize(async (wasm)=>{
    const result = await runTask({
        wasm,
        calls: [
            [
                'AuraApi_slot_duration',
                []
            ]
        ],
        mockSignatureHost: false,
        allowUnresolvedImports: false,
        runtimeLogLevel: 0
    });
    if ('Error' in result) throw new Error(result.Error);
    return (0, _util.u8aToBn)((0, _util.hexToU8a)(result.Call.result).subarray(0, 8 /* u64: 8 bytes */ )).toNumber();
});
const destroyWorker = async ()=>{
    if (!__executor_worker) return;
    const executor = await __executor_worker;
    executor.remote[_comlink.releaseProxy]();
    await new Promise((resolve)=>setTimeout(resolve, 50));
    await executor.terminate();
    __executor_worker = undefined;
};
